[Diagnostics] Check whether missing conformance could be fixed by using `.rawValue`

xedin opened the pull request:
A lot of operators (and most likely regular functions a well) have
overloads that accept i.e. a generic parameter that conforms to
`StringProtocol`, so the best fix in situations when argument is
a raw representable type would be to check whether `.rawValue`
conforms to the expected protocol and use it if so.

Resolves rdar://problem/75367157

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Diagnostics] Check whether missing conformance could be fixed by using `.rawValue`

A lot of operators (and most likely regular functions a well) have
overloads that accept i.e. a generic parameter that conforms to
`StringProtocol`, so the best fix in situations when argument is
a raw representable type would be to check whether `.rawValue`
conforms to the expected protocol and use it if so.

Resolves rdar://problem/75367157

xedin committed:
[Diagnostics] Don't record unwrap fix if `?` was used incorrectly

In situations like `""?` or `42?`, let's not record an unwrap fix
since it was incorrect to use `?` in the first place.

xedin committed:
[Diagnostics] NFC: Use non-initializer diagnostic when raw representable type doesn't conform but `.rawValue` does

xedin commented:
@swift-ci please smoke test

hborla commented on test/ClangImporter/newtype_conformance.swift:
```diff
@@ -16,12 +16,11 @@ func acceptHashable<T: Hashable>(_: T) {}
   acceptEquatable(x)
   acceptHashable(x)
-  acceptComparable(x) // expected-error {{global function 'acceptComparable' requires that 'NSNotification.Name' conform to 'Comparable'}}
+  acceptComparable(x) // expected-error {{type 'NSNotification.Name' does not conform to protocol 'Comparable'}} {{19-19=.rawValue}}
```
Hmm, is it possible to keep the `global function requires that` part, or at least a note pointing to the requirement? 

On second thought, I think it would be nice if the original error stays the same, and add a note saying "did you mean .rawValue" or similar with the fix-it. What do you think?

xedin replied:
Yeah, I think that would make sense, I'd still keep `UseRawValue` fix, to dial with ambiguities, but would diagnose `MissingConformanceFailure` with a tailored note.

xedin replied:
✅

xedin replied:
What do you think?

xedin committed:
[Diagnostics] Even though missing conformance could be fixed by `.rawValue` still diagnose it as conformance issue

Conformance requirement failure is a more descriptive error,
augment it with a note about missing `.rawValue`.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please smoke test Windows platform
